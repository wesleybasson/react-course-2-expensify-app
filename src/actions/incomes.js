import database from '../firebase/firebase';

//ADD INCOME
export const addIncome = (income) => ({
    type: 'ADD_INCOME',
    income
})

export const startAddIncome = (incomeData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = incomeData
        const income = { description, note, amount, createdAt }
        return database.ref(`users/${uid}/incomes`).push(income)
            .then((ref) => {
                dispatch(addIncome({
                    id: ref.key,
                    ...income
                }))
            })
    }
}

//REMOVE INCOME
export const removeIncome = ({ id } = {}) => ({
    type: 'REMOVE_INCOME',
    id
})

export const startRemoveIncome = ({ id } = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/incomes/${id}`).remove().then(() => {
            dispatch(removeIncome({ id }))
        })
    }
}

//EDIT INCOME
export const editIncome = (id, updates) => ({
    type: 'EDIT_INCOME',
    id,
    updates
})

export const startEditIncome = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/incomes/${id}`).update(updates).then(() => {
            dispatch(editIncome(id, updates))
        })
    }
}

//SET INCOMES
export const setIncomes = (incomes) => ({
    type: 'SET_INCOMES',
    incomes
})

export const startSetIncomes = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/incomes`)
        .once('value', (snapshot) => {
            const incomes = []
            snapshot.forEach((childSnapshot) => {
                incomes.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            })
            dispatch(setIncomes(incomes))
        })
    }
}