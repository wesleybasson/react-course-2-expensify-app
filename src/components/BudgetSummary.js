import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpensesTotal from '../selectors/expenses-total';
import selectIncomesTotal from '../selectors/incomes-total';

export class BudgetSummary extends React.Component {
    render() {
        return (
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Budget {this.props.result >= 0 ? 'surplus' : 'deficit'} totals <span>{numeral(this.props.result / 100).format('$0,0.00')}</span></h1>
                </div>
            </div>
        )
    }
}

const mapStateToProps = (state) => ({
    result: selectIncomesTotal(state.incomes) - selectExpensesTotal(state.expenses)
})

export default connect(mapStateToProps)(BudgetSummary)