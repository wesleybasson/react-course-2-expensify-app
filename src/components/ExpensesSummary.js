import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export class ExpensesSummary extends React.Component {
    render() {
        return (
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">Viewing <span>{this.props.number}</span> {this.props.number === 1 ? 'expense' : 'expenses'} totalling <span>{numeral(this.props.total / 100).format('$0,0.00')}</span></h1>
                    {this.props.notshown !== 0 && (
                        <h3 className="page-header__subtitle"><span>{this.props.notshown}</span> {this.props.notshown === 1 ? 'expense' : 'expenses'} not shown.</h3>
                    )}
                    <div className="page-header__actions">
                        <Link className="button" to="/createexpense">Add Expense</Link>
                    </div>
                </div>
            </div>
        );
    };
};

const mapStateToProps = (state) => ({
    number: selectExpenses(state.expenses, state.filters).length,
    total: selectExpensesTotal(selectExpenses(state.expenses, state.filters)),
    notshown: selectExpenses(state.expenses, {
        text: '',
        sortBy: 'date',
        startDate: undefined,
        endDate: undefined
    }).length - selectExpenses(state.expenses, state.filters).length
});

export default connect(mapStateToProps)(ExpensesSummary);