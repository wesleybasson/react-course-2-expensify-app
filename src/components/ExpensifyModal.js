import React from 'react';
import Modal from 'react-modal';

export default (props) => (
    <Modal
        isOpen={props.modalOpen}
        onRequestClose={props.handleClose}
        contentLabel="Expensify"
        closeTimeoutMS={200}
        className="modal"
        ariaHideApp={false}
    >
        <h3 className="modal__title">{props.title}</h3>
        <p className="modal__body">Do you really want to remove this item?</p>
        <div className="container">
            <button className="button button__half button--success" onClick={props.handleAction}>Yes</button>
            <button className="button button__half button--cancel" onClick={props.handleClose}>No</button>
        </div>
    </Modal>
);