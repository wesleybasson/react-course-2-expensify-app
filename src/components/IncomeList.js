import React from 'react';
import { connect } from 'react-redux';
import IncomeListItem from './IncomeListItem';
import selectIncomes from '../selectors/incomes';

export const IncomeList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Incomes</div>
            <div className="show-for-desktop">Income</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.incomes.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No Incomes</span>
                    </div>
                ) : (
                    props.incomes.map((income) => (
                        <IncomeListItem key={income.id} {...income} />
                    ))
                )
            }        
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        incomes: selectIncomes(state.incomes, state.filters)
    };
};

export default connect(mapStateToProps)(IncomeList);
