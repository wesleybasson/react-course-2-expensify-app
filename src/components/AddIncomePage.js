import React from 'react';
import { connect } from 'react-redux';
import IncomeForm from './IncomeForm';
import { startAddIncome } from '../actions/incomes';

export class AddIncomePage extends React.Component {
    onSubmit = (income) => {
        this.props.startAddIncome(income);
        this.props.history.push('/incomes');
    };
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Add Income</h1>
                    </div>
                </div>
                <div className="content-container">
                    <IncomeForm
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddIncome: (income) => dispatch(startAddIncome(income))
});

export default connect(undefined, mapDispatchToProps)(AddIncomePage);