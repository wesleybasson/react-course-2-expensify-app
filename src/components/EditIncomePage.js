import React from 'react';
import { connect } from 'react-redux';
import IncomeForm from './IncomeForm';
import { startEditIncome, startRemoveIncome } from '../actions/incomes';
import RemoveModal from './ExpensifyModal';

export class EditIncomePage extends React.Component {
    state = {
        modalOpen: false
    };
    onSubmit = (income) => {
        this.props.startEditIncome(this.props.income.id, income );
        this.props.history.push('/incomes');
    };
    onRemove = () => {
        this.props.startRemoveIncome({ id: this.props.income.id });
        this.props.history.push('/incomes');
    };
    handleOpen = () => {
        this.setState({ modalOpen: true });
    };
    handleClose = () => {
        this.setState({ modalOpen: false });
    };
    render() {
        const title = "Remove Income";
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                        <h1 className="page-header__title">Edit Income</h1>
                    </div>
                </div>
                <div className="content-container">
                    <IncomeForm
                        income={this.props.income}
                        onSubmit={this.onSubmit}
                    />
                    <button className="button button--secondary" onClick={this.handleOpen}>{title}</button>
                </div>
                <RemoveModal
                    modalOpen={this.state.modalOpen}
                    handleClose={this.handleClose}
                    handleAction={this.onRemove}
                    title={title}
                />
            </div>
        );
    }
}

const mapStateToProps = (state, props) => ({
    income: state.incomes.find((income) => income.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
    startEditIncome: (id, income) => dispatch(startEditIncome(id, income)),
    startRemoveIncome: (id) => dispatch(startRemoveIncome(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditIncomePage);