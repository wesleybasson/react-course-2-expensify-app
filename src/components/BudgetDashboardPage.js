import React from 'react';
import BudgetSummary from './BudgetSummary';

const BudgetDashboardPage = () => (
    <div>
        <BudgetSummary />
    </div>
);

export default BudgetDashboardPage;