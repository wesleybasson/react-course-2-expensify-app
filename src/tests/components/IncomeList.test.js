import React from 'react';
import { shallow } from 'enzyme';
import { IncomeList } from '../../components/IncomeList';
import incomes from '../fixtures/incomes';

test('should render IncomeList with incomes', () => {
    const wrapper = shallow(<IncomeList incomes={incomes} />)
    expect(wrapper).toMatchSnapshot();
});

test('should render IncomeList with empty message', () => {
    const wrapper = shallow(<IncomeList incomes={[]} />)
    expect(wrapper).toMatchSnapshot();
});