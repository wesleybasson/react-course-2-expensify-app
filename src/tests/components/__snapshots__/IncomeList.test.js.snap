// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render IncomeList with empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Incomes
    </div>
    <div
      className="show-for-desktop"
    >
      Income
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No Incomes
      </span>
    </div>
  </div>
</div>
`;

exports[`should render IncomeList with incomes 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Incomes
    </div>
    <div
      className="show-for-desktop"
    >
      Income
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <Component
      amount={200000}
      createdAt={0}
      description="Salary"
      id="1"
      key="1"
      note=""
    />
    <Component
      amount={234500}
      createdAt={-345600000}
      description="Rent Income"
      id="2"
      key="2"
      note=""
    />
    <Component
      amount={10999}
      createdAt={345600000}
      description="Refund"
      id="3"
      key="3"
      note=""
    />
  </div>
</div>
`;
